gprofiler/application_separators.py

```diff
@@ -0,0 +1,232 @@
+import configparser
+import os.path
+import re
+from abc import ABCMeta, abstractmethod
+from typing import List, Optional
+
+from granulate_utils.linux.ns import resolve_host_path
+from psutil import NoSuchProcess, Process
+
+from gprofiler.log import get_logger_adapter
+
+logger = get_logger_adapter(__name__)
+
+
+_PYTHON_BIN_RE = re.compile(r"^python([23](\.\d{1,2})?)?$")
+
+
+def _is_python_bin(bin_name: str):
+    return _PYTHON_BIN_RE.match(os.path.basename(bin_name)) is not None
+
+
+class ApplicationSeparator(metaclass=ABCMeta):
+    @abstractmethod
+    def is_supported(self, process: Process) -> bool:
+        pass
+
+    @abstractmethod
+    def get_application_name(self, process: Process) -> Optional[str]:
+        pass
+
+    @property
+    @abstractmethod
+    def app_prefix(self) -> str:
+        pass
+
+
+_IP_PORT_RE = re.compile(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}(?:\:\d{2,5})?$")  # Matches against (ip(:port))
+
+
+def get_cli_arg_value(args: List[str], arg_name: str, check_for_equals_arg: bool = False) -> Optional[str]:
+    if arg_name in args:
+        return args[args.index(arg_name) + 1]
+
+    if check_for_equals_arg:
+        for arg in args:
+            arg_name, _, arg_val = arg.rpartition("=")
+            if arg_val is not None:
+                return arg_val
+
+    return None
+
+
+def append_python_module_to_proc_wd(process: Process, module_name: str) -> str:
+    return f'{process.cwd().replace("/", ".").strip(".")}.{module_name}'
+
+
+class GunicornApplicationSeparator(ApplicationSeparator):
+    @property
+    def app_prefix(self) -> str:
+        return "gunicorn"
+
+    def is_supported(self, process: Process) -> bool:
+        # Either gunicorn (for example: /usr/local/bin/gunicorn) or python that runs gunicorn
+        # (For example: /usr/local/bin/python /usr/local/bin/gunicorn)
+        if "gunicorn" in process.cmdline()[0]:
+            return True
+        if len(process.cmdline()) >= 2:
+            return "gunicorn" in process.cmdline()[1]
+        return False
+
+    def get_application_name(self, process: Process) -> Optional[str]:
+        # As of gunicorn documentation the WSGI module name most probably will come from the cmdline and not from the
+        # config file / environment variables (they added the option to specify `wsgi_app` only in 20.1.0)
+        for arg in process.cmdline():
+            if ":" in arg:
+                if _IP_PORT_RE.match(arg):
+                    continue
+
+                return append_python_module_to_proc_wd(process, arg)
+
+        logger.warning(f"GunicornApplicationSeparator: matched against process {process} but couldn't find WSGI module")
+        return None
+
+
+class UwsgiApplicationSeparator(ApplicationSeparator):
+    @property
+    def app_prefix(self) -> str:
+        return "uwsgi"
+
+    def is_supported(self, process: Process) -> bool:
+        return "uwsgi" in process.cmdline()[0]
```

Comment:
nope, actually it might be a part of an absolute path `/usr/local/bin/uwsgi` - can do basename but I think this one will suffice

Reply:
You can still
```suggestion
        return process.cmdline()[0] == "uwsgi" or process.cmdline()[0].endswith("/uwsgi")
```
It's better to have false negatives of not identifying applications and add more cases in the future.

Reply:
Agree with @YishaiZinkin on this:

> It's better to have false negatives of not identifying applications and add more cases in the future.

It's better to fail in identification instead of identifying incorrectly.

Anyway, @YishaiZinkin 's suggestion is... equivalent to `basename`? So let's just change it to basename.